impl-endpoints/cloudflare-go/runner.go

```diff
@@ -85,55 +88,92 @@ func (t testCaseDC) ServerConfig() (*tls.Config, error) {
 	return config, nil
 }
 
-func (t testCaseDC) ClientConnectionHandler(conn *tls.Conn) error {
+func (t testCaseDC) ClientConnectionHandler(conn *tls.Conn, err error) error {
+	if err != nil {
+		return fmt.Errorf("Unexpected error: \"%s\"")
+	}
 	if conn.ConnectionState().VerifiedDC == nil {
 		return errors.New("Failed to verify a delegated credential")
 	}
 	return nil
 }
 
-func (t testCaseDC) ServerConnectionHandler(conn *tls.Conn) error {
-	// TODO
-	return errors.New("NOT IMPLEMENTED")
+func (t testCaseDC) ServerConnectionHandler(conn *tls.Conn, err error) error {
+	panic("test case not implemented")
 }
 
-type echTestResult struct {
-	serverStatus tls.EXP_EventECHServerStatus
-}
+// The "ech-accept" test case in which a client offers the ECH extension and
+// the server accepts. The client verifies server name "backend.com". See
+// draft-ietf-tls-esni.
+type testCaseECHAccept struct{}
 
-func (r *echTestResult) eventHandler(event tls.EXP_Event) {
-	switch e := event.(type) {
-	case tls.EXP_EventECHServerStatus:
-		r.serverStatus = e
-	}
+func (t testCaseECHAccept) ClientConfig() (*tls.Config, error) {
+	return echClientConfig("/test-inputs/ech_configs")
 }
 
-type testCaseECHAccept struct{}
+func (t testCaseECHAccept) ServerConfig() (*tls.Config, error) {
+	return echServerConfig("/test-inputs/ech_key")
+}
 
-func (t testCaseECHAccept) ClientConfig() (*tls.Config, error) {
-	base64ECHConfigs, err := ioutil.ReadFile("/test-inputs/ech_configs")
+func (t testCaseECHAccept) connectionHandler(conn *tls.Conn, err error) error {
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("Unexpected error: \"%s\"", err)
 	}
 
-	rawECHConfigs, err := base64.StdEncoding.DecodeString(string(base64ECHConfigs))
-	if err != nil {
-		return nil, err
+	st := conn.ConnectionState()
+	if !st.ECHAccepted {
+		return errors.New("ECH not accepted")
 	}
+	log.Printf("Connection completed with backend server \"%s\"", st.ServerName)
+	return nil
+}
 
-	echConfigs, err := tls.UnmarshalECHConfigs(rawECHConfigs)
-	if err != nil {
-		return nil, err
+func (t testCaseECHAccept) ClientConnectionHandler(conn *tls.Conn, err error) error {
+	return t.connectionHandler(conn, err)
+}
+
+func (t testCaseECHAccept) ServerConnectionHandler(conn *tls.Conn, err error) error {
+	return t.connectionHandler(conn, err)
+}
+
+// The "ech-rejet" test case in which the client offers ECH with stale configs
+// and the server rejects. The client verifies server name "client-facing.com",
+// then aborts the handshake with an "ech_required" alert. See
```

Comment:
Have you tried checking with `validatepcap` that this was actually sent?

Reply:
Not yet. i need to rewrite both "ech-accept" and "ech-reject" test cases to use validpcap. For now, I'll leave this as a TODO.

Reply:
👍 a TODO sounds fine